pex/pyenv.py

```diff
@@ -0,0 +1,211 @@
+                for path_entry in os.environ.get("PATH", "").split(os.pathsep):
+                    pyenv_exe = os.path.join(path_entry, "pyenv")
+                    if is_exe(pyenv_exe):
+                        process = subprocess.Popen(args=[pyenv_exe, "root"], stdout=subprocess.PIPE)
+                        stdout, _ = process.communicate()
+                        if process.returncode == 0:
+                            pyenv_root = str(stdout).strip()
+                            break
+
+            if pyenv_root:
+                pyenv = cls(pyenv_root)
+                TRACER.log("A pyenv installation was found: {}".format(pyenv), V=6)
+                return pyenv
+
+            TRACER.log("No pyenv installation was found.", V=6)
+            return None
+
+    @attr.s(frozen=True)
+    class Shim(object):
+        pyenv = attr.ib()  # type: Pyenv
+        path = attr.ib()  # type: str
+        name = attr.ib()  # type: str
+        major = attr.ib()  # type: Optional[str]
+        minor = attr.ib()  # type: Optional[str]
+
+        _SHIM_REGEX = re.compile(
+            r"""
+            ^
+            (?P<name>
+                python |
+                pypy
+            )
+            (?:
+                # Major version
+                (?P<major>[2-9])
+                (?:
+                    \.
+                    # Minor version
+                    (?P<minor>[0-9])
+                    # Some pyenv pythons include a suffix on the interpreter name, similar to
+                    # PEP-3149. For example, python3.6m to indicate it was built with pymalloc.
+                    [a-z]?
+                )?
+            )?
+            $
+            """,
+            flags=re.VERBOSE,
+        )
+
+        @classmethod
+        def parse(cls, pyenv, binary):
+            # type: (Pyenv, str) -> Optional[Pyenv.Shim]
+            """Parses shim information from a python binary path if it looks like a pyenv shim."""
+            if os.path.dirname(binary) != os.path.join(pyenv.root, "shims"):
+                return None
+            match = cls._SHIM_REGEX.match(os.path.basename(binary))
+            if match is None:
+                return None
+            return cls(
+                pyenv=pyenv,
+                path=binary,
+                name=match.group("name"),
+                major=match.group("major"),
+                minor=match.group("minor"),
+            )
+
+        _PYENV_CPYTHON_VERSION_LEADING_CHARS = frozenset(str(digit) for digit in range(2, 10))
+
+        def select_version(self, search_dir=None):
+            # type: (Optional[str]) -> Optional[str]
+            """Reports the active shim version for the given directory or $PWD.
+
+            If the shim is not activated, returns `None`.
+            """
+            with TRACER.timed("Calculating active version for {}...".format(self), V=6):
+                active_versions = self.pyenv.active_versions(search_dir=search_dir)
```

Comment:
Note: at first I thought this logic was incorrect, because the way  https://github.com/pyenv/pyenv#choosing-the-python-version is phrased makes it seem like the fallthrough logic is per-shim. So for example if shell is 3.7.3 and local is 3.6.6 then `python3.6` should work. But I've verified that in practice it does not, and shell overrides local overrides global for all shims at once, so we're good.

Reply:
Yes. I went through the same tests and read the pyenv source code to verify this.

Reply:
The unit tests cover this as well.